apps/emqx_durable_storage/src/emqx_ds_replication_layer.erl

```diff
@@ -0,0 +1,211 @@
+    case emqx_ds_proto_v1:make_iterator(Node, Shard, StorageStream, TopicFilter, StartTime) of
+        {ok, Iter} ->
+            {ok, #iterator{shard = Shard, enc = Iter}};
+        Err = {error, _} ->
+            Err
+    end.
+
+-spec next(iterator(), pos_integer()) -> emqx_ds:next_result(iterator()).
+next(Iter0, BatchSize) ->
+    #iterator{shard = Shard, enc = StorageIter0} = Iter0,
+    Node = node_of_shard(Shard),
+    %% TODO: iterator can contain information that is useful for
+    %% reconstructing messages sent over the network. For example,
+    %% when we send messages with the learned topic index, we could
+    %% send the static part of topic once, and append it to the
+    %% messages on the receiving node, hence saving some network.
+    %%
+    %% This kind of trickery should be probably done here in the
+    %% replication layer. Or, perhaps, in the logic layer.
+    case emqx_ds_proto_v1:next(Node, Shard, StorageIter0, BatchSize) of
+        {ok, StorageIter, Batch} ->
+            Iter = #iterator{shard = Shard, enc = StorageIter},
+            {ok, Iter, Batch};
+        Other ->
+            Other
+    end.
+
+%%================================================================================
+%% behavior callbacks
+%%================================================================================
+
+%%================================================================================
+%% Internal exports (RPC targets)
+%%================================================================================
+
+-spec do_open_shard_v1(shard_id(), emqx_ds:create_db_opts()) -> ok.
+do_open_shard_v1(Shard, Opts) ->
+    emqx_ds_storage_layer:open_shard(Shard, Opts).
+
+-spec do_drop_shard_v1(shard_id()) -> ok.
+do_drop_shard_v1(Shard) ->
+    emqx_ds_storage_layer:drop_shard(Shard).
+
+-spec do_get_streams_v1(shard_id(), emqx_ds:topic_filter(), emqx_ds:time()) ->
+    [{integer(), _Stream}].
+do_get_streams_v1(Shard, TopicFilter, StartTime) ->
+    emqx_ds_storage_layer:get_streams(Shard, TopicFilter, StartTime).
+
+-spec do_make_iterator_v1(shard_id(), _Stream, emqx_ds:topic_filter(), emqx_ds:time()) ->
+    {ok, iterator()} | {error, _}.
+do_make_iterator_v1(Shard, Stream, TopicFilter, StartTime) ->
+    emqx_ds_storage_layer:make_iterator(Shard, Stream, TopicFilter, StartTime).
+
+-spec do_next_v1(shard_id(), Iter, pos_integer()) -> emqx_ds:next_result(Iter).
+do_next_v1(Shard, Iter, BatchSize) ->
+    emqx_ds_storage_layer:next(Shard, Iter, BatchSize).
+
+%%================================================================================
+%% Internal functions
+%%================================================================================
+
+shard_id(DB, Node) ->
+    %% TODO: don't bake node name into the schema, don't repeat the
+    %% Mnesia's 1M$ mistake.
+    {DB, Node}.
+
+-spec node_of_shard(shard_id()) -> node().
+node_of_shard({_DB, Node}) ->
+    Node.
+
+list_nodes() ->
+    mria:running_nodes().
```

Comment:
Q: This will give a list of core nodes only, right?

Reply:
No, it gives the list of all nodes.

Reply:
True, my bad.